pkg/types/utils.go

```diff
@@ -0,0 +1,41 @@
+/* Copyright © 2021 VMware, Inc. All Rights Reserved.
+   SPDX-License-Identifier: Apache-2.0 */
+
+package types
+
+import (
+	"context"
+
+	appsv1 "k8s.io/api/apps/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// CheckIfNCPK8sResourceExists infers the k8s object from resName and checks
+// if it exists
+func CheckIfNCPK8sResourceExists(
+	c client.Client, resName string) (bool, error) {
+	instance := identifyAndGetInstance(resName)
+	instanceDetails := types.NamespacedName{
+		Namespace: NsxNamespace,
+		Name:      resName,
+	}
+	err := c.Get(context.TODO(), instanceDetails, instance)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func identifyAndGetInstance(resName string) runtime.Object {
```

Comment:
opinion comment: we are encapsulating in this function the logic for leveraging the appropriate resource type.
This simplfies using CheckifNcpResourceExists, but also implies that if we should add any other resource here we will need to modify the func (and you have to find it first).

I suggest either:
1) keep type specification explicit as it was before
2) do the same but using types (NsxNcpBootstrapDsName could become a struct which also specifies it's a ds for instance (this might however require a bit of refactoring)
3) have explicit funcs "idenfiyandGetDeployment" and "identifyAndGetDaemonset"

but after all, this is just an opinion, so feel free to take it or ignore it!

Reply:
I know, I thought about these but then it's unlikely we'll add a new resource and won't go over this path to include its logic. Still, we should future proof it. let me update